File: 17-01-simple-share/java-17-01-simple-share/app/src/main/java/com/example/simpleshare/MainActivity.java
package com.example.simpleshare;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.res.AssetManager;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.ShareActionProvider;
import androidx.core.content.FileProvider;
import androidx.core.view.MenuItemCompat;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class MainActivity extends AppCompatActivity {
    private ShareActionProvider shareActionProvider;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        MenuItem item = menu.findItem(R.id.action_share);
        shareActionProvider = (ShareActionProvider) MenuItemCompat.getActionProvider(item);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission("android.permission.WRITE_EXTERNAL_STORAGE") == PackageManager.PERMISSION_GRANTED) {
                new ImageAsyncTask().execute();
            } else {
                String[] permissions = {
                        "android.permission.WRITE_EXTERNAL_STORAGE"
                };
                int requestCode = 200;
                requestPermissions(permissions, requestCode);
            }
        } else {
            new ImageAsyncTask().execute();
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void copyImageFromAsset() {
        String imagePath = getFilesDir() + File.separator + "cat.jpg";
        if (new File(imagePath).exists()) {
            return;
        }

        AssetManager assetManager = getAssets();
        try {
            InputStream inputStream = assetManager.open("cat.jpg");
            OutputStream outputStream = new FileOutputStream(imagePath);
            byte[] buffer = new byte[inputStream.available()];
            inputStream.read(buffer);
            outputStream.write(buffer);
            inputStream.close();
            outputStream.flush();
            outputStream.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    class ImageAsyncTask extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... voids) {
            copyImageFromAsset();
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            if (shareActionProvider != null) {
                Intent shareIntent = new Intent(Intent.ACTION_SEND);
                shareIntent.setType("image/jpg");
                String imagePath = getFilesDir() + File.separator + "cat.jpg";
                File imageFile = new File(imagePath);
                Uri assetUri = FileProvider.getUriForFile(MainActivity.this, "com.example.simpleshare.fileprovider", imageFile);
                shareIntent.putExtra(Intent.EXTRA_STREAM, assetUri);
                shareActionProvider.setShareIntent(shareIntent);
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == 200 && grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            new ImageAsyncTask().execute();
        }
    }
}

File: 17-01-simple-share/kotlin-17-01-simple-share/app/src/main/java/com/example/simpleshare/MainActivity.kt
package com.example.simpleshare

import android.content.Intent
import android.content.pm.PackageManager
import android.os.AsyncTask
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.ShareActionProvider
import androidx.core.content.FileProvider
import androidx.core.view.MenuItemCompat
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class MainActivity : AppCompatActivity() {
    private var shareActionProvider: ShareActionProvider? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_main, menu)
        val item = menu.findItem(R.id.action_share)
        shareActionProvider = MenuItemCompat.getActionProvider(item) as ShareActionProvider
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission("android.permission.WRITE_EXTERNAL_STORAGE") == PackageManager.PERMISSION_GRANTED) {
                ImageAsyncTask().execute()
            } else {
                val permissions = arrayOf("android.permission.WRITE_EXTERNAL_STORAGE")
                val requestCode = 200
                requestPermissions(permissions, requestCode)
            }
        } else {
            ImageAsyncTask().execute()
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        val id = item.itemId


        return if (id == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)

    }

    private fun copyImageFromAsset() {
        val imagePath = filesDir.toString() + File.separator + "cat.jpg"
        if (File(imagePath).exists()) {
            return
        }

        val assetManager = assets
        try {
            val inputStream = assetManager.open("cat.jpg")
            val outputStream = FileOutputStream(imagePath)
            val buffer = ByteArray(inputStream.available())
            inputStream.read(buffer)
            outputStream.write(buffer)
            inputStream.close()
            outputStream.flush()
            outputStream.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    internal inner class ImageAsyncTask : AsyncTask<Void, Void, Void>() {

        override fun doInBackground(vararg voids: Void): Void? {
            copyImageFromAsset()
            return null
        }

        override fun onPostExecute(aVoid: Void) {
            if (shareActionProvider != null) {
                val shareIntent = Intent(Intent.ACTION_SEND)
                shareIntent.type = "image/jpg"
                val imagePath = filesDir.toString() + File.separator + "cat.jpg"
                val imageFile = File(imagePath)
                val assetUri = FileProvider.getUriForFile(this@MainActivity, "com.example.simpleshare.fileprovider", imageFile)
                shareIntent.putExtra(Intent.EXTRA_STREAM, assetUri)
                shareActionProvider?.setShareIntent(shareIntent)
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == 200 && grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            ImageAsyncTask().execute()
        }
    }
}